### PR TITLE
User hooks

### PR DESCRIPTION
**Pull request by lferraz**
#### Changes

Fixes # (issue)

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: limbus/core/component.py [hunk 1]
```diff
@@ -74,16 +75,22 @@ def __call__(self, state: Optional[ComponentState] = None, msg: Optional[str] =
...
-            if msg is None:
-                log.info(f" {comp_name}({iters}): {state.name}")
+            if self._message is None:
+                log.info(f" {self._component.name}({self._component.counter}): {self._state.name}")
             else:
-                log.info(f" {comp_name}({iters}): {state.name} ({msg})")
+                log.info(f" {self._component.name}({self._component.counter}): {self._state.name} ({self._message})")
+
+    @property
+    def message(self) -> Optional[str]:
```
Comment (edgarriba) [inline limbus/core/component.py]:
```suggestion
    def message(self) -> str | None:
```

File: limbus/core/component.py [hunk 2]
```diff
@@ -154,9 +164,9 @@ def stopping_iteration(self) -> int:
         return self._stopping_iteration
 
     @property
-    def state(self) -> ComponentState:
-        """Get the current state of the component."""
-        return self._state.state
+    def state(self) -> Tuple[ComponentState, Optional[str]]:
```
Comment (edgarriba) [inline limbus/core/component.py]:
```suggestion
    def state(self) -> tuple[ComponentState, str | None]:
```
from future annotations

File: limbus/core/component.py [hunk 3]
```diff
@@ -306,15 +316,15 @@ async def __call__(self) -> None:
...
+        if self.state[0] in [ComponentState.STOPPED, ComponentState.STOPPED_AT_ITER,
+                             ComponentState.ERROR, ComponentState.FORCED_STOP]:
             return True
         return False
 
     def _stop_if_needed(self) -> bool:
         """Stop the component if it is required."""
         if self.is_stopped():
-            if self.state is not ComponentState.STOPPED_AT_ITER:
+            if self.state[0] is not ComponentState.STOPPED_AT_ITER:
```
Comment (edgarriba) [inline limbus/core/component.py]: no guards to check random access index?

File: limbus/core/component.py (hunk 1 above)
Comment (lferraz) [reply limbus/core/component.py]: this does not work for all the supperted python versions.

File: limbus/core/component.py (hunk 3 above)
Comment (lferraz) [reply limbus/core/component.py]: still a draft this PR. XD A lot of work is missing. I do not like this index but it was the easy way to do it.

File: limbus/core/component.py (hunk 1 above)
Comment (edgarriba) [reply limbus/core/component.py]: even with futures ?

Comment (lferraz) [reply limbus/core/component.py]: not sure if this is only supported for pytohn >3.8 or >3.9. Once we remove the support we can start using that notation. Doing it in that way is not giving us any improvement apart of a nicest way to annotate the typings.

**Comment by lferraz** [comment]
NOTE: for compatibility reasons with python 3.8 when declaring params in components we must use the old typing tags. E.g. instead of using `list` we still need to use `List`.

Add a note in some place (maybe readme) denoting this issue.

File: limbus/core/pipeline.py [hunk 4]
```diff
@@ -190,6 +190,9 @@ async def before_component_hook(self, component: Component) -> None:
             component.set_state(ComponentState.PAUSED)
         await self._resume_event.wait()
         component.set_state(ComponentState.READY)
+        # update the iteration counter
+        # since each component can be running a different iteration we assign the max value
+        self._counter = max(self.counter, component.counter)
```
Comment (edgarriba) [inline limbus/core/pipeline.py]: BTW, i had trouble with this variable the other day while creating a counter in a custom Component ? I think is a common name potentially used b users. Can we use internally just the private variable or give it a more precise name to be used internally, something like `_pipelines_iterations_counter` ?

Comment (lferraz) [reply limbus/core/pipeline.py]: maybe we can use double _ for this vars.

Comment (lferraz) [reply limbus/core/pipeline.py]: anyway I don ot like the name, it should be changed.

**Pull request merged**